pkg/storage/legacysql/dualwrite/managed_mode3.go

```diff
@@ -0,0 +1,134 @@
+package dualwrite
+
+import (
+	"context"
+	"net/http"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	grafanarest "github.com/grafana/grafana/pkg/apiserver/rest"
+)
+
+func (m *service) NewStorage(gr schema.GroupResource,
+	legacy grafanarest.LegacyStorage,
+	storage grafanarest.Storage,
+) (grafanarest.Storage, error) {
+	status, _ := m.Status(context.Background(), gr)
+
+	// Support startup modes without special
+	if !status.Runtime {
+		if status.ReadUnified {
+			if status.WriteLegacy {
+				// Write both, read unified
+				return grafanarest.NewDualWriter(grafanarest.Mode3, legacy, storage, m.reg, gr.String()), nil
+			}
+			return storage, nil
+		}
+		if status.WriteUnified {
+			// Write both, read legacy
+			return grafanarest.NewDualWriter(grafanarest.Mode2, legacy, storage, m.reg, gr.String()), nil
+		}
+		return legacy, nil
+	}
+
+	return &mangedMode3{
+		service: m,
+		legacy:  legacy,
+		unified: storage,
+		target:  grafanarest.NewDualWriter(grafanarest.Mode3, legacy, storage, m.reg, gr.String()),
+		gr:      gr,
+	}, nil
+}
+
+type mangedMode3 struct {
```

Comment:
Suggestion: We should add some tests around the semantics of this new mode like we got in https://github.com/grafana/grafana/tree/91242340c18b71696ef598d9c56f71fb6db63c6c/pkg/apiserver/rest

Reply:
💯  I'll do that in a PR that is targeting main -- this is against a feature branch